### PR TITLE
Ignore Sync errors on pipes when doing `CheckAttributeReader.CheckPath`, fix the hang of `git cat-file`

### LINKED ISSUE #17094: CheckAttributeReader.CheckPath hangs forever, and make the background process never end
### Gitea Version

dev(main)

### Git Version

2.30.2

### Operating System

Linux

### How are you running Gitea?

Docker & self-build

### Database

MySQL

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Description

https://github.com/go-gitea/gitea/blob/main/modules/git/repo_attribute.go#L180

CheckAttributeReader.CheckPath hangs forever, and make the background process never end.

Related PR: 
debug helper: https://github.com/go-gitea/gitea/pull/17082
introduced by: https://github.com/go-gitea/gitea/pull/16773

```
1 @ 0x4382d6 0x4309de 0x4634a9 0x4ceef2 0x4d11c5 0x4d11b2 0x4da225 0x4da21b 0x929585 0x93cf6d 0x1b3332d 0x1b33caf 0x146fa62 0x1478263 0x1477437 0x468c81
#	0x4634a8	internal/poll.runtime_pollWait+0x88					/usr/local/go/src/runtime/netpoll.go:229
#	0x4ceef1	internal/poll.(*pollDesc).wait+0x31					/usr/local/go/src/internal/poll/fd_poll_runtime.go:84
#	0x4d11c4	internal/poll.(*pollDesc).waitWrite+0x2c4				/usr/local/go/src/internal/poll/fd_poll_runtime.go:93
#	0x4d11b1	internal/poll.(*FD).Write+0x2b1						/usr/local/go/src/internal/poll/fd_unix.go:283
#	0x4da224	os.(*File).write+0x64							/usr/local/go/src/os/file_posix.go:49
#	0x4da21a	os.(*File).Write+0x5a							/usr/local/go/src/os/file.go:176
#	0x929584	code.gitea.io/gitea/modules/git.(*CheckAttributeReader).CheckPath+0x1a4	/Users/user/work/gitea/modules/git/repo_attribute.go:180
#	0x93cf6c	code.gitea.io/gitea/modules/git.(*Repository).GetLanguageStats+0xdcc	/Users/user/work/gitea/modules/git/repo_language_stats_nogogit.go:109
#	0x1b3332c	code.gitea.io/gitea/modules/indexer/stats.(*DBIndexer).Index+0x4cc	/Users/user/work/gitea/modules/indexer/stats/db.go:55
#	0x1b33cae	code.gitea.io/gitea/modules/indexer/stats.handle+0x8e			/Users/user/work/gitea/modules/indexer/stats/queue.go:23
#	0x146fa61	code.gitea.io/gitea/modules/queue.NewChannelUniqueQueue.func1+0x141	/Users/user/work/gitea/modules/queue/unique_queue_channel.go:71
#	0x1478262	code.gitea.io/gitea/modules/queue.(*WorkerPool).doWork+0x502		/Users/user/work/gitea/modules/queue/workerpool.go:408
#	0x1477436	code.gitea.io/gitea/modules/queue.(*WorkerPool).addWorkers.func1+0x36	/Users/user/work/gitea/modules/queue/workerpool.go:261
```

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by wxiaoguang**
It is not necessary to call Sync on pipes. And the `Sync` usually results in error: "sync |1: invalid argument".

Ref: https://man7.org/linux/man-pages/man2/fdatasync.2.html :

```
EROFS, EINVAL:  fd is bound to a special file (e.g., a pipe, FIFO, or socket) which does not support synchronization.
```

If we return early from `CheckPath` when `Sync` returns error, the `GetLanguageStats` will hang forever (because the stdin's buffer will be fully filled, then `Write` in `CheckPath` blocks)

But I can not confirm the `Sync` behaviors on different platforms, so I think it's OK to keep it while ignore the returned error.

Related issues:

Bug report:
* https://github.com/go-gitea/gitea/issues/17094

Add caller to cat-file batch calls for debug:
* https://github.com/go-gitea/gitea/pull/17089
* https://github.com/go-gitea/gitea/pull/17082

### CODE REVIEW

**Comment by zeripath** [comment]
This was my worry when we added the error check here in the first place. 

But I couldn't force it to happen myself - Did you actually get this error reported? If you're didn't - it's possible something else is actually happening:

I added the `Sync` because I was also seeing an unusual hold-up likely because the `\x00` is not recognised as a line and because the buffer on a pipe can be up to 64Kb.  The `Sync` appeared to cause a flush in my testing so the unexpected hold-up disappeared. 

Now if you're experiencing some block then the `Sync` may not be enough to solve the problem - especially if you're not actually seeing the above error that you were expecting.

**Comment by zeripath** [comment]
Looking again at the code - I think you're likely suffering from the problem I was experiencing earlier. 

Can you repeatedly replicate this issue?

**Comment by zeripath** [comment]
```patch
diff --git a/modules/git/repo_attribute.go b/modules/git/repo_attribute.go
index 0bd7d7e49..f001e65e3 100644
--- a/modules/git/repo_attribute.go
+++ b/modules/git/repo_attribute.go
@@ -112,13 +112,15 @@ func (c *CheckAttributeReader) Init(ctx context.Context) error {
 
 	if len(c.IndexFile) > 0 && CheckGitVersionAtLeast("1.7.8") == nil {
 		cmdArgs = append(cmdArgs, "--cached")
-		c.env = []string{"GIT_INDEX_FILE=" + c.IndexFile}
+		c.env = append(c.env, "GIT_INDEX_FILE="+c.IndexFile)
 	}
 
 	if len(c.WorkTree) > 0 && CheckGitVersionAtLeast("1.7.8") == nil {
-		c.env = []string{"GIT_WORK_TREE=" + c.WorkTree}
+		c.env = append(c.env, "GIT_WORK_TREE="+c.WorkTree)
 	}
 
+	c.env = append(c.env, "GIT_FLUSH=1")
+
 	if len(c.Attributes) > 0 {
 		cmdArgs = append(cmdArgs, c.Attributes...)
 		cmdArgs = append(cmdArgs, "--")
```

Could you check if this helps?

**Comment by wxiaoguang** [comment]
You patch is good, but it doesn't help about this problem.

Linux document has said that a pipe should be `fsync`-ed.

So I think my patch makes the problem better than before.

On you side, the `Sync` is still been called, so you do not worry about buffered data.

On my side, the `Sync` error is ignored, the whole logic is correct. 

If one day we have enough confidence about the pipe's behavior on all platforms, we can optimize this logic again.



ps: I have written a test program to test `Sync`, the result is:

```
2021/09/20 12:44:32 GOOS=darwin
2021/09/20 12:44:32 pipeWriter.Sync error: sync |1: bad file descriptor
```

```
2021/09/20 12:52:25 GOOS=linux
2021/09/20 12:52:25 pipeWriter.Sync error: sync |1: invalid argument
```

It's obviously wrong to call `Sync` on pipes in Linux / macOS.


Test Program:
<details>
```
package main

import (
	"log"
	"os"
	"os/exec"
	"runtime"
)

func main() {
	log.Printf("GOOS=%v\n", runtime.GOOS)

	var err error
	pipeReader, pipeWriter, err := os.Pipe()
	if err != nil {
		log.Fatalf("os.Pipe error: %v", err)
	}

	cmd := exec.Command("/bin/cat")
	cmd.Stdin = pipeReader
	err = cmd.Start()
	if err != nil {
		log.Fatalf("cmd.Start error: %v", err)
	}

	_, err = pipeWriter.WriteString("test")
	if err != nil {
		log.Fatalf("pipeWriter.Write error: %v", err)
	}

	err = pipeWriter.Sync()
	if err != nil {
		log.Fatalf("pipeWriter.Sync error: %v", err)
	}
}
```
</details>

**Comment by wxiaoguang** [comment]
> Did you actually get this error reported? If you didn't - it's possible something else is actually happening:
> Can you repeatedly replicate this issue?

Both yes, I added many debug logs and I am sure that ignore the `Sync` error will make the logic correct, and `CheckPath` can successfully `ReadAttribute` without `Sync` on my side.

And you can refer the test program to test `Sync` on pipes.

**Comment by zeripath** [comment]
I too am running on Linux but didn't get the sync error you do - otherwise I'd have never put it in. I even tested in a Windows virtual machine. 

Hmm, are you building with go 1.17? I know that at some point go moved to use pipe2 - maybe that doesn't error on sync.  Or maybe it's a kernel difference.

---

I also don't understand how an err at the Sync causes a hang. The logic there should abort and cancel the reader before a hang was possible at Read - at least that was how I intended it - so if there's a hang despite the error then that logic is failing somehow.

So I'm doubly confused and that makes me concerned that I've got the channel logic wrong somewhere in there - but I just can't see it.

---

Anyway if GIT_FLUSH env reliably prevents this hang along with ignoring the error on sync then I am happy.

But if the actual slow down/hang is not occurring because of the sync but at read - then you're likely suffering the same issue that the original native git implementation of commit info was suffering. That is, that although the write has been sent it's not reaching the git program or the git program's output is not reaching the read. If GIT_FLUSH prevents this then that was due to git buffering, if not then...

Perhaps placing a runtime.Gosched() in between the write and the read is needed? Another is changing the code to make things non-synchronous - that is push all of the paths in to git attribute channel, closing the writer at the end and reading them all. This would both fill the buffer and force go to schedule the other goroutines.

**Review by zeripath — APPROVED** [review]

**Comment by wxiaoguang** [comment]
### About environment

I have test the `Sync` on pipes with:

* Debian 10 Linux 4.19.0-16-amd64 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
  go version go1.16.2 linux/amd64
* Debian 11 Linux 5.10.0-8-amd64 SMP Debian 5.10.46-4 (2021-08-03) x86_64 GNU/Linux
  go version go1.17 darwin/amd64 with GOOS=linux
* macOS 11.5.2
  go version go1.17 darwin/amd64

`Sync` on pipes always fails and reports error `sync |1: bad file descriptor` for macOS or `sync |1: invalid argument` for Linux.

I believe that since Linux document says that `fsync` on pipes will fail, then here is a problem.

### Why there is a hang

It seems that there are other flaws with `git cat-file` or `git check-attr`, eg: when the git process is killed by timeout, the go routine still exists, then I still can see a background git process has been running for over 3 hours (in the `process.Manager`), but I can not find the process in system's process list by `ps`.

Because there may be unknown bugs, when `Sync` returns an error and `CheckPath` returns early, there are more calls to `CheckPath` and fill the stdin buffer, then the `Write` in `CheckPath` blocks.

File: modules/indexer/stats/db.go [hunk 1]
```diff
@@ -54,10 +54,17 @@ func (db *DBIndexer) Index(id int64) error {
...
 		return err
 	}
-	return repo.UpdateLanguageStats(commitID, stats)
+	err = repo.UpdateLanguageStats(commitID, stats)
+	if err != nil {
+		log.Error("Unable to update language stats for ID %s for default branch %s in %s. Error: %v", commitID, repo.DefaultBranch, repo.RepoPath(), err)
+		return err
+	}
+
+	log.Info("DBIndexer completed language stats for ID %s for default branch %s in %s. stats count: %d", commitID, repo.DefaultBranch, repo.RepoPath(), len(stats))
```
Comment (lafriks) [inline modules/indexer/stats/db.go]: This should probably be at `Debug` level:
```suggestion
	log.Debug("DBIndexer completed language stats for ID %s for default branch %s in %s. stats count: %d", commitID, repo.DefaultBranch, repo.RepoPath(), len(stats))
```

Comment (wxiaoguang) [reply modules/indexer/stats/db.go]: IMO:
Info for function level, eg: we print logs for route handler, here is a DBIndexer task, it is like a route handler.
Debug for detailed trace or internal data.

Debug is also OK here, applied.

**Review by 6543 — APPROVED** [review]

File: modules/git/repo_language_stats_gogit.go [hunk 2]
```diff
@@ -99,7 +107,7 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 				if language, has := attrs["linguist-language"]; has && language != "unspecified" && language != "" {
 					// group languages, such as Pug -> HTML; SCSS -> CSS
 					group := enry.GetLanguageGroup(language)
-					if len(group) == 0 {
+					if len(group) != 0 {
```
Comment (lunny) [inline modules/git/repo_language_stats_gogit.go:110]: Maybe some parts of this could be backport to v1.15

Comment (zeripath) [reply modules/git/repo_language_stats_gogit.go:110]: 1.15 is fine - this was an error I added in the previous pr.

**Pull request merged**